### PR TITLE
634 implement python bindings for abm overhaul

### LINKED ISSUE #634: Implement python bindings for ABM overhaul
### Feature description

After the merge of #327 , the python bindings need to be updated.

### Additional context

_No response_

### Checklist

- [X] Attached labels, especially loc:: or model:: labels.
- [X] Linked to project.

### LINKED ISSUE #638: LocationID in ABM
### Motivation / Current Behaviour

Currently we have a LocationID in Location_type.h and also a Location has a Type and Index Member, that's convoluting and not clear when we use what.

### Enhancement description

Merge them into one object

### Additional context

_No response_

### Checklist

- [X] Attached labels, especially loc:: or model:: labels.
- [X] Linked to project

### PR DESCRIPTION
**Pull request by xsaschako**
## Changes

- changed locationID in Location.h instead of having index and type seperately
- changed the savingsof locations in world to have ine vektor instead of several
- fixed python bindings
- fixed a sanitizer error which has been caused by integer overflow

## Merge Request - GuideLine Checklist

- *Check our [git workflow](https://github.com/DLR-SC/memilio/wiki/git-workflow) before opening a Pull request/Merge request.*
- *Request a reviewer when your work is ready to review, before this please use the draft feature.*

### Checks by code author

- [x] Every addressed issue is linked (use the "Closes #ISSUE" keyword below)
- [x] New code adheres to [coding guidelines](https://github.com/DLR-SC/memilio/wiki/Coding-guidelines)
- [x] No large data files have been added (files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] Tests are added for new functionality and a local test run was successful
- [x] Appropriate **documentation** for new functionality has been added (Doxygen in the code and Markdown files if necessary)
- [x] Proper attention to licenses, especially no new third-party software with conflicting license has been added

### Checks by code reviewer(s)

- [ ] Corresponding issue(s) is/are linked and addressed
- [ ] Code is clean of development artifacts (no deactivated or commented code lines, no debugging printouts, etc.)
- [ ] Appropriate **unit tests** have been added, CI passes and code coverage is acceptable (did not decrease)
- [ ] No large data files added in the whole history of commits(files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)

### Additional context

closes #638
closes #634

### CODE REVIEW

File: pycode/memilio-simulation/memilio/simulation/abm.cpp [hunk 1]
```diff
@@ -200,9 +187,10 @@ PYBIND11_MODULE(_simulation_abm, m)
     py::class_<mio::abm::World>(m, "World")
         .def(py::init<mio::abm::GlobalInfectionParameters>(),
              py::arg("infection_parameters") = mio::abm::GlobalInfectionParameters{})
-        .def("add_location", &mio::abm::World::add_location, py::arg("location_type"), py::arg("num_cells") = 0,
+        .def("add_location", &mio::abm::World::add_location, py::arg("location_type"), py::arg("num_cells") = 1,
+             py::return_value_policy::reference_internal)
```
Comment (mknaranja) [inline pycode/memilio-simulation/memilio/simulation/abm.cpp:193]: Is this correct? `add_location ` does not return a reference (in contrast to `add_person`)

**Review by mknaranja — COMMENTED** [review]
Thanks for quickly tacking the pybind issue @xsaschako 

I only had two comments but as I only looked at it briefly, please also have a look @DavidKerkmann

File: cpp/models/abm/person.cpp [hunk 2]
```diff
@@ -110,6 +110,10 @@ const Location& Person::get_location() const
 
 void Person::set_assigned_location(Location& location)
 {
+    /* TODO: This is not safe if the location is not the same as added in the world, e.g. the index is wrong. We need to check this.
```
[the remark this answers is not in the document]
Comment (xsaschako) [reply cpp/models/abm/person.cpp:113]: Very true.

File: pycode/memilio-simulation/memilio/simulation/abm.cpp (hunk 1 above)
Comment (xsaschako) [reply pycode/memilio-simulation/memilio/simulation/abm.cpp:193]: Good Call!

File: pycode/memilio-simulation/memilio/simulation/abm.cpp [hunk 3]
```diff
@@ -170,12 +164,7 @@ PYBIND11_MODULE(_simulation_abm, m)
 
     py::class_<mio::abm::Location>(m, "Location")
         .def_property_readonly("type", &mio::abm::Location::get_type)
-        .def_property_readonly("index", &mio::abm::Location::get_index)
-        .def_property("infection_parameters",
-                      py::overload_cast<>(&mio::abm::Location::get_infection_parameters, py::const_),
-                      [](mio::abm::Location& self, mio::abm::LocalInfectionParameters params) {
-                          self.get_infection_parameters() = params;
-                      });
```
Comment (DavidKerkmann) [inline pycode/memilio-simulation/memilio/simulation/abm.cpp:178]: Why was this removed? Local infection parameters should not have changed.

File: pycode/memilio-simulation/memilio/simulation_test/test_abm.py [hunk 4]
```diff
@@ -42,14 +42,11 @@ def test_locations(self):
...
+        self.assertEqual(len(world.locations), 2)
 
-        home = world.locations[home_id.type][home_id.index]
+        home = world.locations[home_id.index]
         self.assertEqual(home.type, abm.LocationType.Home)
 
-        home.infection_parameters.MaximumContacts = 10
-        self.assertEqual(home.infection_parameters.MaximumContacts, 10)
-
         testing_ages = [abm.AgeGroup.Age0to4]
```
Comment (DavidKerkmann) [inline pycode/memilio-simulation/memilio/simulation_test/test_abm.py:50]: This should also work then I think.

**Review by DavidKerkmann — COMMENTED** [review]
Only two remarks.
We should decide if we want to track people and locations by reference or by some index/id and use the same structure for both.

**Review by mknaranja — APPROVED** [review]
